crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -2775,64 +2784,118 @@ impl Forks {
-                    if fork.markers.is_disjoint(&markers) {
+                for fork in std::mem::take(&mut forks) {
+                    let Some((remaining_forker, envs)) = forker.fork(&fork.env, conflicting_groups)
+                    else {
                         new.push(fork);
                         continue;
-                    }
+                    };
+                    forker = remaining_forker;
 
-                    let not_markers = markers.negate();
-                    let mut new_markers = markers.clone();
-                    new_markers.and(fork.markers.negate());
-                    if !fork.markers.is_disjoint(&not_markers) {
+                    for fork_env in envs {
                         let mut new_fork = fork.clone();
-                        new_fork.intersect(not_markers);
+                        new_fork.set_env(fork_env);
+                        // We only add the dependency to this fork if it
+                        // satisfies the fork's markers. Some forks are
+                        // specifically created to exclude this dependency,
+                        // so this isn't always true!
+                        if forker.included(&new_fork.env) {
+                            new_fork.add_dependency(dep.clone());
+                        }
                         // Filter out any forks we created that are disjoint with our
                         // Python requirement.
-                        if !new_fork.markers.is_disjoint(&python_marker) {
+                        if new_fork.env.included_by_marker(&python_marker) {
                             new.push(new_fork);
                         }
                     }
-                    fork.dependencies.push(dep.clone());
-                    fork.intersect(markers);
-                    // Filter out any forks we created that are disjoint with our
-                    // Python requirement.
-                    if !fork.markers.is_disjoint(&python_marker) {
-                        new.push(fork);
-                    }
-                    markers = new_markers;
                 }
                 forks = new;
             }
         }
+        // When there is a conflicting group configuration, we need
+        // to potentially add more forks. Each fork added contains an
+        // exclusion list of conflicting groups where dependencies with
+        // the corresponding package and extra name are forcefully
+        // excluded from that group.
+        //
+        // We specifically iterate on conflicting groups and
+        // potentially re-generate all forks for each one. We do it
+        // this way in case there are multiple sets of conflicting
+        // groups that impact the forks here.
+        //
+        // For example, if we have conflicting groups {x1, x2} and {x3,
+        // x4}, we need to make sure the forks generated from one set
+        // also account for the other set.
```

Comment:
This could potentially lead to a lot of forks, but... so be it? Unavoidable.

Reply:
Yup. I don't see another path here.

And note that the way I went about this _should_ hopefully minimize forks if possible. For example, one way of dealing with conflicting extras/groups is to use it to create an initial set of forks initially in resolution. And then you don't have to do it here: you just need to rely on forks not including dependencies in the extra/group exclusions. But doing it this way means that _every_ marker-related fork gets duplicated according to the declared conflicts. But with this approach, _only_ the forks that contain a conflict end up getting multiplied. So if there's an optional dependency that only exists in one marker fork and there are ten marker forks, the conflicting exclusions only fork off of that one fork and not all of them. (If that makes sense.)

But it's not immediately obvious to me how to reduce this further or if it's possible.